### ISSUE TITLE
Consistency of TaskSpec._run_hook() return value

### ISSUE THREAD

**Issue by asauve2** [sartography/SpiffWorkflow#336]
Following the discussion in #334, I noticed that there is a different behaviour 
with the task state transition when `TaskSpec._run_hook()`returns `False`.
We need  solid understanding of this point for clean implementation of threading initiated 
from `run_hook()`.

The different behaviour for each case are:
1. Normal work : `return True`
    * transition observed:  `WAITING => COMPLETED`
2. Task requiring to complete things later: `return False` 
    * transition observed `WAITING => RUNNING` # this transition is very welcome for us
3. Error event processed with `task.workflow.catch(ErrorEventDefinition(bpmn_id)); return False`  
    * transition observed `WAITING => CANCELLED`

Is case 2 correct and stable for next releases (as it is a core design feature for implementing threading) ?

Case 3 seems inconsistent with case 2.  
If the use case has been correctly written above, this may be improved with an enumeration
or an explicit call somewhere as the implicit translation of the boolean value may lead to
code difficult to handle.

**Comment by essweine** [comment]
When `_run_hook` 
 - returns `True`: task transitions to `COMPLETED`
 - returns `False`: task transitions to `ERROR`
 - returns `None`: task transitions to `STARTED`

The transition to `CANCELLED` can happen any time.  An interrupting boundary event cancels the task it's attached to if the event is caught, so that's what's driving the transition in your case.

The boundary event does introduce some inconsistency (overriding the `ERROR` state with `CANCELLED`).  But the BPMN spec says that an interrupting boundary event cancels the task it's attached to, so that's the reason.  You could change the boundary event type to non-interrupting and then the task would not get cancelled and should move to `ERROR` if the method returns `False`, and the boundary event will still execute.

`STARTED` is equivalent to running,   At this point, Spiff can go on to other branches, which the branch that the `STARTED` task is on waits.

The piece currently not implemented automatically is handling transitions out of `STARTED`, which your workflow runner would have to do.  The reason for not implementing something like `poll` was that it wasn't entirely clear to me how it should work given that it's something new and I didn't want to force everyone to live with it a potentially bad decision.  So I thought it might be better to just add the state and then figure out how to use it (I do want to add it).  This functionality is somewhat experimental at this point (though the state changing, because there should be an option to execute a task asynchronously, the question is only how to manage).

This was more or less what I was thinking of (but this isn't fully baked either):
https://github.com/sartography/spiff-example-cli/blob/threaded-service-task/threaded_service.py

This isn't exactly like your use case, as you aren't using the script engine to execute the service tasks, but spawning threads could just as easily happen in a custom task, and the workflow runner could manage them directly (it has to explicitly handle the `STARTED` tasks now anyway).  In that case, it would be handling catching and sending the error back to the workflow (I didn't add a `try` when retrieving the futures, but you could), and would either call `task.complete` or `task.error` depending on the result.

Not sure if that example or this explanation is helpful, so let me know if you have questions.

**Comment by asauve2** [comment]
Thanks for these precision @essweine now, after coming back to the code in the right places,
all makes sense.


When trying to make things explicit using constants for return values of _run_hook,
it appeared that returning the *desired* next `TaskState` seems to be the most consistent.

It may look like:
```
def _run_hook(self, my_task):
    try:
        if delayed_result:
            return TaskState.STARTED
        else:
            return TaskState.COMPLETED
    except:
        return TaskState.ERROR
```

Obvious error in chosen `TaskState` should probably raise exception in `Task`,
but this way it leaves room for future return values states extension.

It leaves the fact that the `Task` will make the state change in the above case
and some other state change can be initiated from the code itself, which looks
like dispersion of state handling. But still returning explicit `TaskState` seems
to help understanding the process.

What do you think about that?

I'm currently working on threading itself and will come back with results later, 
but main choice boils down to initiate thread runners at workflow level or
at spec level. The later looking to be more in line with actual library design.

**Comment by essweine** [comment]
The main reason for `True`/`False`/`None` was to make the method operate similarly to `_update` which returns `True` or `False` according to whether the task should more to `READY` vs `WAITING`.  But it would be reasonable to return a task state there as well.  I think overall this is a good idea and will discuss with the team.

**Comment by asauve2** [comment]
Finally we got this treaded implementation running!

The thread model looks similar to what you linked in  https://github.com/sartography/spiff-example-cli/blob/threaded-service-task/threaded_service.py
Still here are the crunchy points to make sure we got the same understanding:
- thread is submitted from `taskSpec._run_hook()` using a dedicated threaded task wrapper
- TaskState change must be initiated at top level workflow runner level, not from thread to avoid race conditions 
  (or in other words, state transitions happens only in main thread)

The pseudo code of main runner looks like
- while not `BmfWorkflow.is_completed()`
    - `BmfWorkflow.do_engine_steps()`
    - update `TaskState` of finished threads from scheduled state stored in threaded wrapper

All of this looks currently healthy but our validation phase is not completed.

Coming back to the the `_run_hook()` return value.
In the implementation phase it appeared clearly that it would make sense to add a precision to previous proposal:
- return `TaskState` for synchronized applications (as proposed above)
- return `None` for asynchronous applications to request NO state transition and delegate explicitely TaskChange to upper level workflow

Here the `None`  return is very different from actual v2.0.0rc2 implementation but 
allow smoother implementations for async and synchronous applications.
